### DISCUSSION TITLE
Access to the defaults arguments of methods and ParamTermClause

### DISCUSSION THREAD

**Discussion by rssh** [scala/scala3#14078] [Metaprogramming]
Currently, It's no way to find default arguments of function from macro.
Logically,  DefDef have TermParamClause wich provide List(Valdef) and default arguments can be there.
But currently,  all rhs are empty.  I.e. next code:
```
 defdef.termParamss.headOption match
                 case Some(paramClause) =>
                   paramClause match
                     case TermParamClause(valDefs) =>
                         val paramsWithDefauls = valDefs.filter(_.rhs.isDefined).map(vd => (vd.symbol.name, vd.rhs.get.toString))
                         Expr.ofList(paramsWithDefaults.map(Expr(_)))
                     case _ =>  throw new RuntimeException("TermParamClause is expected")
                 case None =>  ???

```
 Always return an empty list for all defdefs.

From one side, this can be a bug, from other --- semantics of default-values is other, maybe better way is to provide some other API for this, so this should be a feature request?

Note, that in Scala2, some code generation libraries use default values of a primary constructor (in a quite hacky way, for example 
 - https://github.com/plokhotnyuk/jsoniter-scala/blob/b89681a8838ca2c9c760ad84be4021811eef4d91/jsoniter-scala-macros/shared/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/macros/JsonCodecMaker.scala#L939. )
 -

**Comment by som-snytt** [comment]
Besides forward- and backward-compatible, perhaps there is also hackward-compatible.

**Reply by plokhotnyuk** [reply]
@som-snytt Please take a look at it from the user's side. This feature is required by different serialization/transformation libraries and wasn't implemented in Scala 3 macro API yet. 

[A lot of Scala libraries depend on it](https://github.com/search?l=Scala&q=%22%24lessinit%24greater%24default%24%22&type=Code)

**Reply by som-snytt** [reply]
I think I was just making a pun, not a value judgment, because of "in a hacky way".

I'm sure I recently saw an issue about default args with a comment that it must wait until future changes to how default args are implemented.

I know from some experience that default args are fragile in scala 2 even from the compiler side.

**Comment by rssh** [comment]
Wow, in dottty. $lessinit$greater$default also works:  https://github.com/lampepfl/dotty-macro-examples/blob/master/defaultParamsInference/src/macro.scala.   (was in search)

**Reply by rssh** [reply]
(after some thinking -- the original question [is the absence of default values in. term clause is bug or not] is still a question. And access to defaultParams should be documented somewhere]
